src/bin/upgrade/main.rs

```diff
@@ -258,34 +253,39 @@ fn get_dependencies(
         .collect::<Result<BTreeMap<_, _>>>()?;
 
     let mut upgrades = DesiredUpgrades::default();
-    for dependency in package
-        .dependencies
-        .clone()
+    for (dependency, old_version) in manifest
+        .get_dependencies()
+        .map(|(_, result)| result.map_err(Error::from))
+        .collect::<Result<Vec<_>>>()?
         .into_iter()
-        .filter(is_version_dep)
-        .filter(|dependency| !exclude.contains(&dependency.name))
+        .filter_map(|dependency| {
+            dependency
+                .version()
+                .map(ToOwned::to_owned)
+                .map(|version| (dependency, version))
+        })
+        .filter(|(dependency, _)| !exclude.contains(&dependency.name))
         // Exclude renamed dependencies as well
-        .filter(|dependency| {
+        .filter(|(dependency, _)| {
             dependency
-                .rename
-                .as_ref()
-                .map_or(true, |rename| !exclude.contains(rename))
+                .rename()
+                .map_or(true, |rename| !exclude.iter().any(|s| s == rename))
         })
     {
-        let is_prerelease = dependency.req.to_string().contains('-');
+        let registry = dependency
+            .registry()
+            .map(|registry| registry_url(&manifest.path, Some(registry)))
+            .transpose()?;
```

Comment:
How come we are now doing the registry lookup at this stage?

Reply:
Because the code seemed to assume that `UpgradeMetadata`'s `registry` field should hold a url.

I don't really have an opinion about this. However, if we switched to storing registry names in `UpgradeMetadata` instead of urls, then I think the lookup would have to occur in two places instead of one (like it is now).

Reply:
Oh, right. Ideally, we'd rename that to clarify it but that isn't the focus of this PR, so I can understand not doing it.